compiler/src/main/kotlin/io/spine/protodata/QueryBuilder.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2021, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.protodata
+
+import com.google.common.collect.Iterables.getOnlyElement
+import io.grpc.stub.StreamObserver
+import io.spine.base.EntityState
+import io.spine.base.Identifier.checkSupported
+import io.spine.client.ActorRequestFactory
+import io.spine.client.Query
+import io.spine.client.QueryResponse
+import io.spine.core.UserId
+import io.spine.protobuf.AnyPacker
+import io.spine.server.BoundedContext
+import java.util.*
+
+/**
+ * A builder for queries to the projections defined on top of the Protobuf compiler events.
+ */
+public class QueryBuilder<T : EntityState>
+internal constructor(
+    private val context: BoundedContext,
+    private val type: Class<T>,
+    actorName: String
+) {
+
+    private val actor = UserId.newBuilder()
+        .setValue(actorName)
+        .build()
+    private val factory = ActorRequestFactory.newBuilder()
+        .setActor(actor)
+        .build()
+
+    /**
+     * Selects a projection by its ID.
+     */
+    public fun withId(id: Any): SingleCastQuery<T> {
+        checkSupported(id.javaClass)
+        return SingleCastQuery(context, buildQuery(id), type)
+    }
+
+    /**
+     * Selects all projections of the given type.
+     */
+    public fun all(): MulticastQuery<T> {
+        return MulticastQuery(context, buildQuery(), type)
+    }
+
+    private fun buildQuery(id: Any? = null): Query {
+        val queries = factory.query()
+        return if (id == null) {
+            queries.all(type)
+        } else {
+            queries.byIds(type, setOf(id))
+        }
+    }
+}
+
+/**
+ * A query which may yield many entities.
+ */
+public class MulticastQuery<T : EntityState>(
```

Comment:
I think there is a confusion between Multicast-Singlecast queries and multicast-unicast routing we already have in Spine.

To me, multicast is more about transmitting data between equal peers, rather than about the power of the result set  in a method call.

I would have just one Query type, with `execute() : Set<T>` and `findOne(): T` methods. It's either way a developer's decision somewhere — be it on the Query implementation or on the method call. But with "one type-two methods" system (c) we'll have one type less to support.